src/Nethermind/Nethermind.Merge.Plugin/InvalidChainTracker/InvalidChainTracker.cs

```diff
@@ -0,0 +1,188 @@
+//  Copyright (c) 2021 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+// 
+
+using System;
+using System.Collections.Generic;
+using Nethermind.Blockchain;
+using Nethermind.Blockchain.Find;
+using Nethermind.Consensus;
+using Nethermind.Consensus.Processing;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+using Nethermind.Merge.Plugin.Handlers;
+
+namespace Nethermind.Merge.Plugin.InvalidChainTracker;
+
+/// <summary>
+/// Something that tracks if a given hash is on a known invalid chain, as one if it's ancestor have been reported to
+/// be invalid.
+/// 
+/// </summary>
+public class InvalidChainTracker: IInvalidChainTracker
+{
+    private readonly IPoSSwitcher _poSSwitcher;
+    private readonly IBlockFinder _blockFinder;
+    private readonly IBlockCacheService _blockCacheService;
+    private readonly ILogger _logger;
+    private readonly object _opLock = new();
+    private readonly LruCache<Keccak, Node> _tree;
+    
+    // CompositeDisposable only available on System.Reactive. So this will do for now. 
+    private readonly List<Action> _disposables = new();
+
+    public InvalidChainTracker(
+        IPoSSwitcher poSSwitcher, 
+        IBlockFinder blockFinder,
+        IBlockCacheService blockCacheService,
+        ILogManager logManager)
+    {
+        _poSSwitcher = poSSwitcher;
+        _blockFinder = blockFinder;
+        _tree = new LruCache<Keccak, Node>(1024, nameof(InvalidChainTracker));
+        _logger = logManager.GetClassLogger<InvalidChainTracker>();
+        _blockCacheService = blockCacheService;
+    }
+
+    public void SetupBlockchainProcessorInterceptor(IBlockchainProcessor blockchainProcessor)
+    {
+        blockchainProcessor.OnInvalidBlock += BlockchainProcessorOnOnInvalidBlock;
+        _disposables.Add(() =>
+        {
+            blockchainProcessor.OnInvalidBlock -= BlockchainProcessorOnOnInvalidBlock;
+        });
+    }
+
+    private void BlockchainProcessorOnOnInvalidBlock(object? sender, InvalidBlockException e)
+    {
+        OnInvalidBlock(e.InvalidBlockHash, null);
+    }
```

Comment:
If we made it EventArgs I think we can easily pass ParentHash

Reply:
InvalidBlockException does not contain the parent hash, and I don't want to mess around too much.

Reply:
You can easily add it to OnInvalidBlockArg now.